interview-questions.md

```diff
@@ -10,40 +10,102 @@ Then, research the question to expand on your answer. Even if you feel you have
 
 Your answer:
 
-Researched answer:
+The earliest computers relied entirely upon functional programming. In essence, these Turing machines operated on data by executing sequential lines of code to perform basic operations. Early computers used punch cards to read instructions for calculations, reading one line after another. More advanced computers began to introduce techniques for using conditional statements to determine whether to perform an operation, loops to iterate over commands, and tags to jump between different segments of code. Modern functional code consists largely of functions that are called or invoked within an application. These functions receive arguments, process through a statement, and can return data. Functions provide reuseable and repeatable behavior.
+
+Much like functional programming, object-oriented programming (OOP) seeks to provide reusable and repeatable behavior. Abstraction takes similar types of data and functions and combines them into classes. Classes define the types of data that they contain and provide access to their data through methods. During execution, objects are initialized as an instance of a class. OOP seeks to provide data security through an idea known as encapsulation. Each object protects its data (or attributes) from direct access, instead providing methods to define specific ways for creating, retrieving, updating, or deleting data. Many languages also implement public and private attributes and methods to further protect data. 
+
+Object-oriented programming (OOP) and functional programming are often viewed as opposing approaches to coding. Opponents of OOP often point to the challenges associated with effectively defining useful and meaningful classes and with teaching OOP. In practice, meaningful applications frequently integrate both approaches at varying points. While objects can provide useful ways to visualize and solve problems, they complicate others. It's important to consider both approaches and make meaningful decisions in when to create a function vs when to create a class and implement objects.
+
+Researched answer: I started going down a rabbit-hole looking into the relative utility of both paradigms within parallel and distributed computing problems. One table comparing the relative strengths and weakness of these options indicated that OOP does not support parallel programming. This seemed like an interesting and strong assertion. Looking deeper, I think that there is a more nuanced explanation that OOP requires some deliberate organization of a system, allowing subcomponents and their objects to function independently of other subcomponents, whereby both subcomponents can complete their calculations independently. Functional programming seems to be much more commonly used in academic environments, particularly by scientists. This is likely a confluence of perpetual education, system efficiency, and a preference for highly-logical (and sometimes linear) styles of thought.
```

Comment:
🤯 WOW!  This may be the most thorough initial answer I have seen!  The way you talk about these concepts are so in-depth and well articulated that it seems like it is coming straight out of documentation somewhere (not implying anything because I am pretty confident this is in your own words).  Outstanding work!